### ISSUE TITLE
link C docs from R docs

### ISSUE THREAD

**Issue by maelle** [igraph/rigraph#1317]
in each manual page.

Find a way to make that automatic.

cc @szhorvat

**Comment by maelle** [comment]
It would not need to be automatic.

We could add a special tag automatically for already exposed functions, and manually for new exposed functions.

**Comment by maelle** [comment]
for each function that is defined as or wraps an `_impl` function we can find what is `.Call` inside it.

Now from say `R_igraph_empty` we can know we want to link to `igraph_empty`. How do we find the link to https://igraph.org/c/doc/igraph-Basic.html#igraph_empty?

We could parse the docs folder of the C library every time we update the vendored file. In each xxml file there are links like

https://github.com/igraph/igraph/blob/03760a09cc7195dfaad3a1e60cff1632aa6a6118/doc/basicigraph.xxml#L176

so for each xxml file we can know what C things are documented in it.

- the URL fragment is the name of the C thing e.g. `#igraph_empty`
- the base URL is `https://igraph.org/c/doc/`
- the path is given by the id of the chapter in the xxml file https://github.com/igraph/igraph/blob/03760a09cc7195dfaad3a1e60cff1632aa6a6118/doc/basicigraph.xxml#L7

**Comment by maelle** [comment]
so the tag we'd add would be something like

```
#'  @cmethod igraph_empty
```

and we'd use an internal lookup table to know in which chapter of the C docs it is documented.

**Comment by szhorvat** [comment]
> How do we find the link to https://igraph.org/c/doc/igraph-Basic.html#igraph_empty?

You can parse the index, https://igraph.org/c/html/latest/ix01.html

The name of the target to build docs for C locally is `html`.

> so for each xxml file we can know what C things are documented in it.

Trying to get this from xxml files isn't going to be reliable.

----

Be careful, some R functions map to multiple C functions! Example `distances()`.

**Comment by maelle** [comment]
then on how we make the special tag work I'll mimic https://github.com/moodymudskipper/devtag

**Comment by maelle** [comment]
> Be careful, some R functions map to multiple C functions! Example distances().

Ah then they'd need multiple tags.

Great to know reg https://igraph.org/c/html/latest/ix01.html thank you!

**Comment by maelle** [comment]
starting the R package over at https://github.com/igraph/igraph.r2cdocs, I parsed the index in https://github.com/igraph/igraph.r2cdocs/blob/main/data-raw/c-links.R

**Comment by maelle** [comment]
https://roxygen2.r-lib.org/articles/extending.html

**Comment by maelle** [comment]
actually it won't need to be a separate package it will live inside rigraph

**Comment by maelle** [comment]
https://stackoverflow.com/questions/77865500/how-can-i-create-custom-roxygen2-tags-for-a-package

**Comment by maelle** [comment]
@szhorvat I don't see the several igraph methods `distances()` is calling?

**Comment by szhorvat** [comment]
> @szhorvat I don't see the several igraph methods `distances()` is calling?

This is why I suggested to skip these special cases for now. The interface is often sufficiently different that the C docs are less relevant. That said, you can see it like this: 

Look up that `distances()` calls `R_igraph_shortest_paths()`:

https://github.com/igraph/rigraph/blob/5298fe8f6ba1958b57a638583fe4420ec79c3caa/R/structural.properties.R#L902

This does NOT mean that this maps to `igraph_shortest_paths()`. We have to find where `R_igraph_shortest_paths()` comes from. In this case, it is not auto-generated, so it's found in `rinterface_extra.c`:

https://github.com/igraph/rigraph/blob/5298fe8f6ba1958b57a638583fe4420ec79c3caa/src/rinterface_extra.c#L4186

Looking at the sources, you'll see that currently it makes use of several functions such as `igraph_distances`, `igraph_distances_dijkstra`, `igraph_distances_bellman_ford`, `distances_johnson`, `igraph_distances_floyd_warshall`, and even calls the locally defined `distances_johnson`, which in turn calls others: https://github.com/igraph/rigraph/blob/5298fe8f6ba1958b57a638583fe4420ec79c3caa/src/rinterface_extra.c#L4159

If you see the `_impl`, that means full auto-generation, so predictability.

**Comment by maelle** [comment]
For a function that `.Call()`s `R_igraph_square_lattice` how do I know what igraph functions `R_igraph_square_lattice` calls? Here's what I think I understand:

- If I find "igraph_square_lattice" in `tools/stimulus/functions.yaml` (or `tools/stimulus/functions-R.yaml`, which one to use?) it means the igraph function really is "igraph_square_lattice". Or can I simply use the index of the C docs? If there's a `.Call(R_something)` and igraph has a function called `something` then probably it's the one?

- If not it means I have to parse `src/rinterface_extra.c`. I should try and get better at treesitter then :sweat_smile: (which in R only has the grammar for R https://github.com/DavisVaughan/r-tree-sitter/issues/21

**Comment by maelle** [comment]
But if I start by functions that are/wrap `_impl` functions then it's easy, as you said.

**Comment by maelle** [comment]
Regarding `src/rinterface_extra.c` what we need is for each function defined inside it, to know what igraph function it calls **but limited to those igraph functions that are in igraph manual**. For instance, a function that calculates the size of a vector is not relevant for our purposes.

**Comment by schochastics** [comment]
@maelle I feel like this is done or am I wrong?

**Comment by maelle** [comment]
Yes, how could I forget to close the issue? Thanks so much for letting me click this button 😸
